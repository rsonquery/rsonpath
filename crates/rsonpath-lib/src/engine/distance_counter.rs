use lazy_static::lazy_static;
use log::debug;
use std::collections::HashMap;
use std::fs::File;
use std::io::{BufWriter, Write};
use std::path::Path;
use std::sync::Mutex;

// Track how many times every distance appears in the skipping process, only track the ones with frequency >= 1
lazy_static! {
    static ref DISTANCE_COUNTER_MAP: Mutex<HashMap<usize, usize>> = Mutex::new(HashMap::new());
}

// Increment the value by 1, or initialize to 1 if not present
pub fn increment_value(distance: usize) {
    let mut map = DISTANCE_COUNTER_MAP.lock().unwrap();
    *map.entry(distance).or_insert(0) += 1;
}

// Save HashMap to a CSV file
pub fn save_to_csv(file_path: &str) -> std::io::Result<()> {
    debug!("Saving to {}", file_path);
    let map = DISTANCE_COUNTER_MAP.lock().unwrap();
    let path = Path::new(file_path);
    let file = File::create(path)?;
    let mut writer = BufWriter::new(file);

    for (key, value) in map.iter() {
        debug!("Writing {}, {}", key, value);
        writeln!(writer, "{},{}", key, value)?;
    }

    Ok(())
}
