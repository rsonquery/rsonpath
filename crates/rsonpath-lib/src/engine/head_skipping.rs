//! Engine decorator that performs **head skipping** &ndash; an extremely optimized search for
//! the first matching member name in a query starting with a self-looping state.
//! This happens in queries starting with a descendant selector.

use std::marker::PhantomData;
use crate::{
    automaton::{Automaton, State},
    classification::{
        mask::Mask,
        memmem::Memmem,
        quotes::{InnerIter, QuoteClassifiedIterator, ResumedQuoteClassifier},
        simd::{dispatch_simd, Simd},
        structural::{BracketType, Structural, StructuralIterator},
        ResumeClassifierBlockState, ResumeClassifierState,
    },
    debug,
    depth::Depth,
    engine::EngineError,
    input::{
        error::{InputError, InputErrorConvertible},
        Input, InputBlockIterator,
    },
    result::Recorder,
    FallibleIterator, MaskType, BLOCK_SIZE,
};
use rsonpath_syntax::str::JsonString;
use crate::result::InputRecorder;

/// Trait that needs to be implemented by an [`Engine`](`super::Engine`) to use this submodule.
pub(super) trait CanHeadSkip<'i, 'r, I, R, V>
where
    I: Input<'i, 'r, R, BLOCK_SIZE> + 'i,
    R: Recorder<I::Block> + 'r,
    V: Simd,
{
    /// Function called when head-skipping finds a member name at which normal query execution
    /// should resume.
    ///
    /// The [`HeadSkip::run_head_skipping`] function will call this implementation
    /// whenever it finds a member name matching the first transition in the query.
    /// The structural `classifier` passed is guaranteed to have classified the
    /// `next_event` and nothing past that. It is guaranteed that
    /// `next_event` is [`Structural::Opening`].
    ///
    /// When called, the engine must start with in the automaton state as given in `state`
    /// and execute the query until a matching [`Structural::Closing`] character is encountered,
    /// using `classifier` for classification and `result` for reporting query results. The `classifier`
    /// must *not* be used to classify anything past the matching [`Structural::Closing`] character.
    fn run_on_subtree(
        &mut self,
        next_event: Structural,
        state: State,
        structural_classifier: V::StructuralClassifier<'i, I::BlockIterator>,
    ) -> Result<ResumeState<'i, I::BlockIterator, V, MaskType>, EngineError>;

    fn recorder(&mut self) -> &'r R;
}

pub(super) struct ResumeState<'i, I, V, M>(
    pub(super) ResumeClassifierState<'i, I, V::QuotesClassifier<'i, I>, M, BLOCK_SIZE>,
)
where
    I: InputBlockIterator<'i, BLOCK_SIZE>,
    V: Simd;

/// Configuration of the head-skipping decorator.
pub(super) struct HeadSkip<'b, 'q, 'r, I, R, V, const N: usize>
where
    I: Input<'b, 'r, R, N> + 'b,
    R: InputRecorder<I::Block> + 'r,
{
    bytes: &'b I,
    state: State,
    is_accepting: bool,
    member_name: &'q JsonString,
    simd: V,
    _recorder: PhantomData<&'r R>,
}

impl<'b, 'q, 'r, I, R, V> HeadSkip<'b, 'q, 'r, I, R, V, BLOCK_SIZE>
where
    I: Input<'b, 'r, R, BLOCK_SIZE>,
    R: Recorder<I::Block>,
    V: Simd
{
    /// Create a new instance of the head-skipping decorator over a given input
    /// and for a compiled query [`Automaton`].
    ///
    /// # Returns
    /// If head-skipping is possible for the query represented by `automaton`,
    /// returns [`Some`] with a configured instance of [`HeadSkip`].
    /// If head-skipping is not possible, returns [`None`].
    ///
    /// ## Details
    /// Head-skipping is possible if the query automaton starts
    /// with a state with a wildcard self-loop and a single member-labelled transition forward.
    /// Syntactically, if the [`fallback_state`](`crate::query::automaton::StateTable::fallback_state`)
    /// of the [`initial_state`](`crate::query::automaton::StateTable::initial_state`) is the same as the
    /// [`initial_state`](`crate::query::automaton::StateTable::initial_state`), and its
    /// [`transitions`](`crate::query::automaton::StateTable::transitions`) are a single-element list.
    ///
    /// This means that we can search for the label of the forward transition in the entire document,
    /// disregarding any additional structure &ndash; during execution we would always loop
    /// around in the initial state until encountering the desired member name. This search can be done
    /// extremely quickly with [`classification::memmem`](crate::classification::memmem).
    ///
    /// In all other cases, head-skipping is not supported.
    pub(super) fn new(bytes: &'b I, automaton: &'b Automaton<'q>, simd: V) -> Option<Self> {
        let initial_state = automaton.initial_state();
        let fallback_state = automaton[initial_state].fallback_state();
        let transitions = automaton[initial_state].member_transitions();

        if fallback_state == initial_state
            && transitions.len() == 1
            && automaton[initial_state].array_transitions().is_empty()
        {
            let (member_name, target_state) = transitions[0];
            debug!("Automaton starts with a descendant search, using memmem heuristic.");
            return Some(Self {
                bytes,
                state: target_state,
                is_accepting: automaton.is_accepting(target_state),
                member_name,
                simd,
                _recorder: PhantomData,
            });
        }

        None
    }

    /// Run a preconfigured [`HeadSkip`] using the given `engine` and reporting
    /// to the `result`.
    pub(super) fn run_head_skipping<E>(&self, engine: &mut E) -> Result<(), EngineError>
    where
        'b: 'r,
        E: CanHeadSkip<'b, 'r, I, R, V>,
        R: Recorder<I::Block> + 'r,
    {
        dispatch_simd!(self.simd; self, engine =>
        fn<'b, 'q, 'r, I, V, E, R>(head_skip: &HeadSkip<'b, 'q, 'r, I, R, V, BLOCK_SIZE>, engine: &mut E) -> Result<(), EngineError>
        where
            'b: 'r,
            E: CanHeadSkip<'b, 'r, I, R, V>,
            R: Recorder<I::Block> + 'r,
            I: Input<'b, 'r, R, BLOCK_SIZE>,
            V: Simd
        {
            let mut input_iter = head_skip.bytes.iter_blocks(engine.recorder());
            let mut idx = 0;
            let mut first_block = None;

            loop {
                let mut memmem = head_skip.simd.memmem(head_skip.bytes, &mut input_iter);
                debug!("Starting memmem search from {idx}");

                if let Some((starting_quote_idx, last_block)) = memmem.find_label(first_block, idx, head_skip.member_name)? {
                    drop(memmem);

                    first_block = Some(last_block);
                    idx = starting_quote_idx;
                    debug!("Needle found at {idx}");
                    let seek_start_idx = idx + head_skip.member_name.quoted().len();

                match head_skip.bytes.seek_non_whitespace_forward(seek_start_idx).e()? {
                    Some((colon_idx, b':')) => {
                        let (next_idx, next_c) = head_skip
                            .bytes
                            .seek_non_whitespace_forward(colon_idx + 1).e()?
                            .ok_or(EngineError::MissingItem())?;

                            let ResumedQuoteClassifier {
                                classifier: quote_classifier,
                                first_block: quote_classified_first_block,
                            } = head_skip.simd.resume_quote_classification(input_iter, first_block);

                            // Temporarily set the index within the current block to zero.
                            // This makes sense for the move below.
                            let mut classifier_state = ResumeClassifierState {
                                iter: quote_classifier,
                                block: quote_classified_first_block
                                    .map(|b| ResumeClassifierBlockState { block: b, idx: 0 }),
                                are_colons_on: false,
                                are_commas_on: head_skip.is_accepting,
                            };

                            debug!("Actual match with colon at {colon_idx}");
                            debug!("Next significant character at {next_idx}");
                            debug!("Classifier is at {}", classifier_state.get_idx());
                            debug!("We will forward to {colon_idx} first, then to {next_idx}",);

                            // Now we want to move the entire iterator state so that the current block is quote-classified,
                            // and correctly points to the place the engine would expect had it found the matching key
                            // in the regular loop. If the value is atomic, we handle it ourselves. If the value is complex,
                            // the engine wants to start one byte *after* the opening character. However, the match report
                            // has to happen before we advance one more byte, or else the opening character might be lost
                            // in the output (if it happens at a block boundary).
                            if next_c == b'{' || next_c == b'[' {
                                forward_to(&mut classifier_state, next_idx)?;
                                if head_skip.is_accepting {
                                    engine.recorder().record_match(
                                        next_idx,
                                        Depth::ZERO,
                                        crate::result::MatchedNodeType::Complex,
                                    )?;
                                }
                                forward_to(&mut classifier_state, next_idx + 1)?;
                            } else {
                                forward_to(&mut classifier_state, next_idx)?;
                            };

                            // We now have the block where we want and we ran quote classification, but during the `forward_to`
                            // call we lose all the flow-through quote information that usually is passed from one block to the next.
                            // We need to manually verify the soundness of the classification. Fortunately:
                            // 1. we know that resume_idx is either the start of a value, or one byte after an opening -
                            //    in a valid JSON this character can be within quotes if and only if it is itself a quote;
                            // 2. the only way the mask can be wrong is if it is flipped - marks chars within quotes
                            //    as outside and vice versa - so it suffices to flip it if it is wrong.
                            if let Some(block) = classifier_state.block.as_mut() {
                                let should_be_quoted = block.block.block[block.idx] == b'"';
                                if block.block.within_quotes_mask.is_lit(block.idx) != should_be_quoted {
                                    debug!("Mask needs flipping!");
                                    block.block.within_quotes_mask = !block.block.within_quotes_mask;
                                    classifier_state.iter.flip_quotes_bit();
                                }
                            }

                            classifier_state = match next_c {
                                b'{' | b'[' => {
                                    debug!("resuming");
                                    let classifier = head_skip.simd.resume_structural_classification(classifier_state);
                                    engine
                                        .run_on_subtree(
                                            Structural::Opening(
                                                if next_c == b'{' {
                                                    BracketType::Curly
                                                } else {
                                                    BracketType::Square
                                                },
                                                next_idx,
                                            ),
                                            head_skip.state,
                                            classifier,
                                        )?
                                        .0
                                }
                                _ if head_skip.is_accepting => {
                                    engine.recorder().record_match(
                                        next_idx,
                                        Depth::ZERO,
                                        crate::result::MatchedNodeType::Atomic,
                                    )?;
                                    let mut classifier = head_skip.simd.resume_structural_classification(classifier_state);
                                    let next_structural = classifier.next()?;

                                    match next_structural {
                                        Some(s) => engine.recorder().record_value_terminator(s.idx(), Depth::ZERO)?,
                                        None => return Err(EngineError::MissingClosingCharacter()),
                                    }
                                    classifier.stop()
                                }
                                _ => classifier_state,
                            };

                            debug!("Quote classified up to {}", classifier_state.get_idx());
                            idx = classifier_state.get_idx();

                            first_block = classifier_state.block.map(|b| b.block.block);
                            input_iter = classifier_state.iter.into_inner();
                        }
                        _ => idx += 1,
                    }
                } else {
                    debug!("No memmem matches, exiting");
                    break;
                }
            }

            return Ok(());

            /// Move the state forward to `index`.
            ///
            /// # Errors
            /// If the offset crosses block boundaries, then a new block is read from the underlying
            /// [`Input`](crate::input::Input) implementation, which can fail.
            ///
            /// # Panics
            /// If the `index` is not ahead of the current position of the state ([`get_idx`](ResumeClassifierState::get_idx)).
            #[inline(always)]
            #[allow(clippy::panic_in_result_fn)]
            fn forward_to<'i, I, Q, M, const N: usize>(state: &mut ResumeClassifierState<'i, I, Q, M, N>, index: usize) -> Result<(), InputError>
            where
                I: InputBlockIterator<'i, N>,
                Q: QuoteClassifiedIterator<'i, I, M, N>,
            {
                let current_block_start = state.iter.get_offset();
                let current_block_idx = state.block.as_ref().map_or(0, |b| b.idx);
                let current_idx = current_block_start + current_block_idx;

                debug!(
                    "Calling forward_to({index}) when the inner iter offset is {current_block_start} and block idx is {current_block_idx:?}"
                );

                // We want to move by this much forward, and delta > 0.
                assert!(index > current_idx);
                let delta = index - current_idx;

                // First we virtually pretend to move *backward*, setting the index of the current block to zero,
                // and adjust the delta to cover that distance. This makes calculations simpler.
                // Then we need to skip zero or more blocks and set our self.block to the last one we visit.
                let remaining = delta + current_block_idx;
                let blocks_to_skip = remaining / N;
                let remainder = remaining % N;

                match state.block.as_mut() {
                    Some(b) if blocks_to_skip == 0 => {
                        b.idx = remaining;
                    }
                    Some(_) => {
                        state.block = state
                            .iter
                            .offset(blocks_to_skip as isize)?
                            .map(|b| ResumeClassifierBlockState {
                                block: b,
                                idx: remainder,
                            });
                    }
                    None => {
                        state.block = state
                            .iter
                            .offset((blocks_to_skip + 1) as isize)?
                            .map(|b| ResumeClassifierBlockState {
                                block: b,
                                idx: remainder,
                            });
                    }
                }

                debug!("forward_to({index}) results in idx moved to {}", state.get_idx());

                Ok(())
            }
        })
    }
}
