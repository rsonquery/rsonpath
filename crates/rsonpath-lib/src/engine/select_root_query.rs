//! Special case handlers for the empty query $.
//!
//! The main engine is built with a path automaton in mind, and for simplicity
//! we assume the root opening was already read. This makes it incompatible with
//! an empty query. Instead of rewriting the engine we provide fast-path implementations
//! here.
use crate::{
    engine::{error::EngineError, Input},
    input::{error::InputErrorConvertible, InputBlockIterator},
    is_json_whitespace,
    result::{empty::EmptyRecorder, Match, MatchCount, MatchIndex, MatchSpan, Sink},
    BLOCK_SIZE,
};
use crate::result::InputRecorder;

/// Count for an empty query &ndash; determine if the root exists.
pub(super) fn count<'i, 'r, I, R, const N: usize>(input: &I) -> Result<MatchCount, EngineError>
where
    I: Input<'i, 'r, R, N>,
    R: InputRecorder<I::Block> + 'r
{
    // Assuming a correct JSON, there is either one root if any non-whitespace character
    // occurs in the document, or the document is empty.
    if input.seek_non_whitespace_forward(0).e()?.is_some() {
        Ok(1)
    } else {
        Ok(0)
    }
}

/// Index for an empty query &ndash; determine the first index of the root.
pub(super) fn index<'i, 'r, I, R, S, const N: usize>(input: &I, sink: &mut S) -> Result<(), EngineError>
where
    I: Input<'i, 'r, R, N>,
    R: InputRecorder<I::Block> + 'r,
    S: Sink<MatchIndex>,
{
    // Assuming a correct JSON, the root starts at the first non-whitespace character, if any.
    if let Some((first_idx, _)) = input.seek_non_whitespace_forward(0).e()? {
        sink.add_match(first_idx - input.leading_padding_len())
            .map_err(|err| EngineError::SinkError(Box::new(err)))?;
    }

    Ok(())
}

/// Approximate span for an empty query &ndash; determine the first index and the length of the root.
pub(super) fn approx_span<'i, 'r, I, R, S, const N: usize>(input: &I, sink: &mut S) -> Result<(), EngineError>
where
    I: Input<'i, 'r, R, N>,
    R: InputRecorder<I::Block> + 'r,
    S: Sink<MatchSpan>,
{
    // The root spans the entire document, by definition, with the exception of whitespace.
    // We need to find the start index exactly, and then can return the length of the rest as the approximate
    // length of the root.
    //
    // Some input know their lengths: bytes already in memory, file mmaps, etc.
    // A BufferedInput over an arbitrary Read stream cannot know its length, so we actually
    // need to iterate until the end and count the bytes.
    if let Some((first_idx, _)) = input.seek_non_whitespace_forward(0).e()? {
        let end_idx = match input.len_hint() {
            Some(end_idx) => end_idx, // Known length, just take it.
            None => {
                // Unknown length, iterate and count.
                let mut iter = input.iter_blocks(&EmptyRecorder);
                let mut end_idx = 0;

                while (iter.next().e()?).is_some() {
                    end_idx += BLOCK_SIZE;
                }

                end_idx
            }
        };

        sink.add_match(MatchSpan::from_indices(
            first_idx - input.leading_padding_len(),
            end_idx - input.leading_padding_len(),
        ))
        .map_err(|err| EngineError::SinkError(Box::new(err)))?;
    }

    Ok(())
}

/// Match for an empty query &ndash; copy the entire document, trimming whitespace.
pub(super) fn match_<'i, 'r, I, R, S, const N: usize>(input: &I, sink: &mut S) -> Result<(), EngineError>
where
    I: Input<'i, 'r, R, N>,
    R: InputRecorder<I::Block> + 'r,
    S: Sink<Match>,
{
    // For a full match we need to copy the entire input starting from first non-whitespace,
    // and then trim the whitespace from the end. This might be slow if the document is excessively
    // padded with whitespace at start and/or end, but that's a pathological case.
    let mut iter = input.iter_blocks(&EmptyRecorder);
    let mut res: Vec<u8> = vec![];
    let mut first_significant_idx = None;
    let mut offset = 0;

    while let Some(block) = iter.next().e()? {
        if first_significant_idx.is_none() {
            // Start of the root not found yet, look for it.
            first_significant_idx = block.iter().position(|&x| !is_json_whitespace(x));

            if let Some(first_idx) = first_significant_idx {
                // Start of the root found in this block, copy the relevant part.
                res.extend(&block[first_idx..]);
            } else {
                offset += block.len();
            }
        } else {
            // Start of the root was already found, now we are copying everything.
            res.extend(&*block);
        }
    }

    if let Some(start) = first_significant_idx {
        // Trim whitespace if we have a result.
        while !res.is_empty() && is_json_whitespace(res[res.len() - 1]) {
            res.pop();
        }

        let actual_start = start + offset - input.leading_padding_len();
        sink.add_match(Match::from_start_and_bytes(actual_start, res))
            .map_err(|err| EngineError::SinkError(Box::new(err)))?;
    }

    Ok(())
}
