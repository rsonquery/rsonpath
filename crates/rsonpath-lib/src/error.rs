//! Common errors shared across the library.
use std::fmt::{self, Display};
use thiserror::Error;

pub(crate) const FEATURE_REQUEST_URL: &str =
    "https://github.com/V0ldek/rsonpath/issues/new?template=feature_request.md";
pub(crate) const BUG_REPORT_URL: &str = "https://github.com/V0ldek/rsonpath/issues/new?template=bug_report.md";

/// Internal irrecoverable error. These are caused solely
/// by bugs &ndash; broken invariants or assertions in internal logic &ndash;
/// but we return those instead of panicking.
#[derive(Error, Debug)]
pub struct InternalRsonpathError {
    details: &'static str,
    #[source]
    source: Option<InternalErrorSource>,
}

/// Errors in internal depth tracking of execution engines.
#[derive(Error, Debug)]
pub enum DepthError {
    /// The engine's maximum depth limit was exceeded.
    /// The inner [`usize`] indicates that limit.
    #[error("Maximum depth of {0} exceeded.")]
    AboveLimit(usize),
    /// The document has unmatched closing characters
    /// and is malformed.
    #[error("Depth fell below zero.")]
    BelowZero,
}

struct InternalErrorSource(Box<dyn std::error::Error + Send + Sync>);

impl fmt::Debug for InternalErrorSource {
    #[inline(always)]
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(&self.0, f)
    }
}

impl Display for InternalErrorSource {
    #[inline(always)]
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Display::fmt(&self.0, f)
    }
}

impl std::error::Error for InternalErrorSource {
    #[inline(always)]
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        self.0.source()
    }
}

impl InternalRsonpathError {
    #[allow(unused)]
    pub(crate) fn from_expectation(details: &'static str) -> Self {
        Self { details, source: None }
    }

    #[allow(unused)]
    pub(crate) fn from_error<E: std::error::Error + Send + Sync + 'static>(err: E, details: &'static str) -> Self {
        Self {
            details,
            source: Some(InternalErrorSource(Box::new(err))),
        }
    }
}

impl Display for InternalRsonpathError {
    #[inline]
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let details_are_enabled = std::env::var("RUST_BACKTRACE").unwrap_or_default() == "0";
        write!(
            f,
            "an internal error has occurred; this is a bug, please report it at {BUG_REPORT_URL}"
        )?;

        if details_are_enabled {
            writeln!(f, "; the error details follow")?;
            write!(f, "{}", self.details)?;
            if let Some(source) = &self.source {
                write!(f, "; source: {source}")?;
            }
        }

        Ok(())
    }
}

/// Error raised when rsonpath is asked to perform an operation that is currently
/// unsupported. This may be either because the feature is in the works, or
/// because it is not planned to ever be supported.
#[derive(Error, Debug)]
pub struct UnsupportedFeatureError {
    issue: Option<usize>,
    feature: &'static str,
}

impl UnsupportedFeatureError {
    #[must_use]
    #[allow(dead_code)]
    #[inline(always)]
    fn tracked(issue: usize, feature: &'static str) -> Self {
        Self {
            issue: Some(issue),
            feature,
        }
    }

    #[must_use]
    #[inline(always)]
    fn untracked(feature: &'static str) -> Self {
        Self { issue: None, feature }
    }

    /// Large JSON Depths feature &ndash; supporting JSON documents
    /// with nesting depth exceeding 255. Unsupported and not planned.
    #[must_use]
    #[inline(always)]
    pub fn large_json_depths() -> Self {
        Self::untracked("Large JSON Depths")
    }

    /// Large Automaton Queries feature &ndash; supporting queries that
    /// cause compiled DFAs to exceed 256 states. Unsupported and not planned.
    #[must_use]
    #[inline(always)]
    pub fn large_automaton_queries() -> Self {
        Self::untracked("Large Automaton Queries")
    }

    /// Multiple Selector per Segment feature &ndash; supporting queries
    /// that contain a union of selectors per step. Unsupported and not planned (yet).
    #[must_use]
    #[inline(always)]
    pub fn multiple_selectors() -> Self {
        Self::untracked("Multiple Selector per Segment")
    }

    /// Indexing from End &ndash; supporting index and slice selectors that
    /// use from-end indexing. Unsupported and not planned (yet).
    #[must_use]
    #[inline(always)]
    pub fn indexing_from_end() -> Self {
        Self::untracked("Indexing from End")
    }

    /// Slice Selector &ndash; supporting slice selectors.
    /// <https://github.com/V0ldek/rsonpath/issues/152>
    #[must_use]
    #[inline(always)]
    pub fn slice_selector() -> Self {
        Self::tracked(152, "Slice Selector")
    }

    /// Slice with Backward Step &ndash; supporting slice selectors that step backwards.
    /// Unsupported and not planned (yet).
    #[must_use]
    #[inline(always)]
    pub fn slice_with_backward_step() -> Self {
        Self::untracked("Slice with Backward Step")
    }

    /// Filter Selector &ndash; supporting filter selectors.
    /// <https://github.com/V0ldek/rsonpath/issues/154>
    #[must_use]
    #[inline(always)]
    pub fn filter_selector() -> Self {
        Self::tracked(154, "Filter Selector")
    }

    /// Returns the issue number on GitHub corresponding to the unsupported feature.
    /// Is [`None`] if the feature is not planned.
    #[must_use]
    #[inline(always)]
    pub fn issue(&self) -> Option<usize> {
        self.issue
    }

    /// Returns the descriptive name of the feature.
    #[must_use]
    #[inline(always)]
    pub fn feature(&self) -> &str {
        self.feature
    }

    /// Whether the issue is planned to ever be supported.
    #[must_use]
    #[inline(always)]
    pub fn is_planned(&self) -> bool {
        self.issue.is_some()
    }
}

impl Display for UnsupportedFeatureError {
    #[inline]
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self.issue {
            Some(issue) => {
                write!(
                    f,
                    "the feature {} (#{}) is not supported yet; it is being tracked and discussed at \
                    https://github.com/V0ldek/rsonpath/issues/{}",
                    self.feature, issue, issue
                )
            }
            None => {
                write!(
                    f,
                    "the feature {} is not supported, and is not planned; \
                    if you would like to see it introduced to rsonpath, please raise a feature request at \
                    {FEATURE_REQUEST_URL}",
                    self.feature
                )
            }
        }
    }
}
