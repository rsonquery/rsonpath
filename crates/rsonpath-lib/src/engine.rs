//! Base traits for different implementations of JSONPath execution engines.
//!
//! Defines the [`Engine`] trait that provides different ways of retrieving
//! query results from input bytes, as well as [`Compiler`] which provides
//! a standalone entry point for compiling a [`JsonPathQuery`] into an [`Engine`].
pub mod error;
mod head_skipping;
pub mod main;
mod tail_skipping;
pub use main::MainEngine as RsonpathEngine;
mod select_root_query;

use self::error::EngineError;
use crate::input::SeekableBackwardsInput;
use crate::{
    automaton::{error::CompilerError, Automaton},
    input::Input,
    result::{Match, MatchCount, MatchIndex, MatchSpan, Sink},
};
use rsonpath_syntax::JsonPathQuery;

/// An engine that can run its query on a given input.
pub trait Engine {
    /// Find the number of matches on the given [`Input`].
    ///
    /// The result is equivalent to using [`matches`](Engine::matches) and counting the matches,
    /// but in general is much more time and memory efficient.
    ///
    /// # Errors
    /// An appropriate [`EngineError`] is returned if the JSON input is malformed
    /// and the syntax error is detected.
    ///
    /// **Please note** that detecting malformed JSONs is not guaranteed.
    /// Some glaring errors like mismatched braces or double quotes are raised,
    /// but in general **the result of an engine run on an invalid JSON is undefined**.
    /// It _is_ guaranteed that the computation terminates and does not panic.
    fn count<I>(&self, input: &I) -> Result<MatchCount, EngineError>
    where
        I: SeekableBackwardsInput;

    /// Find the starting indices of matches on the given [`Input`] and write them to the [`Sink`].
    ///
    /// The result is equivalent to using [`matches`](Engine::matches) and extracting the
    /// [`Match::span.start_idx`],
    /// but in general is much more time and memory efficient.
    ///
    /// # Errors
    /// An appropriate [`EngineError`] is returned if the JSON input is malformed
    /// and the syntax error is detected.
    ///
    /// **Please note** that detecting malformed JSONs is not guaranteed.
    /// Some glaring errors like mismatched braces or double quotes are raised,
    /// but in general **the result of an engine run on an invalid JSON is undefined**.
    /// It _is_ guaranteed that the computation terminates and does not panic.
    fn indices<I, S>(&self, input: &I, sink: &mut S) -> Result<(), EngineError>
    where
        I: SeekableBackwardsInput,
        S: Sink<MatchIndex>;

    /// Find the approximate spans of matches on the given [`Input`] and write them to the [`Sink`].
    ///
    /// "Approximate" means that the ends of spans are not guaranteed to be located exactly at the end of a match,
    /// but may include trailing whitespace. **Importantly**, it may be beyond the rigidly-defined length of the
    /// input, as the engine is allowed to pad the input with whitespace to help with processing. For the purposes
    /// of this API, it is assumed that every character after the logical end of the input is a whitespace character.
    /// With that in mind, it is guaranteed that:
    /// 1. the span start is exact;
    /// 2. the span encompasses the entire matched value;
    /// 3. the only characters included after the value are JSON whitespace characters:
    ///    space (0x20), horizontal tab (0x09), new line (0x0A), carriage return (0x0D).
    ///
    /// # Errors
    /// An appropriate [`EngineError`] is returned if the JSON input is malformed
    /// and the syntax error is detected.
    ///
    /// **Please note** that detecting malformed JSONs is not guaranteed.
    /// Some glaring errors like mismatched braces or double quotes are raised,
    /// but in general **the result of an engine run on an invalid JSON is undefined**.
    /// It _is_ guaranteed that the computation terminates and does not panic.
    fn approximate_spans<I, S>(&self, input: &I, sink: &mut S) -> Result<(), EngineError>
    where
        I: SeekableBackwardsInput,
        S: Sink<MatchSpan>;

    /// Find all matches on the given [`Input`] and write them to the [`Sink`].
    ///
    /// # Errors
    /// An appropriate [`EngineError`] is returned if the JSON input is malformed
    /// and the syntax error is detected.
    ///
    /// **Please note** that detecting malformed JSONs is not guaranteed.
    /// Some glaring errors like mismatched braces or double quotes are raised,
    /// but in general **the result of an engine run on an invalid JSON is undefined**.
    /// It _is_ guaranteed that the computation terminates and does not panic.
    fn matches<I, S>(&self, input: &I, sink: &mut S) -> Result<(), EngineError>
    where
        I: SeekableBackwardsInput,
        S: Sink<Match>;
}

/// An engine that can be created by compiling a [`JsonPathQuery`].
pub trait Compiler {
    /// Concrete type of the [`Engines`](`Engine`) created,
    /// parameterized with the lifetime of the input query.
    type E<'q>: Engine + 'q;

    /// Compile a [`JsonPathQuery`] into an [`Engine`].c
    ///
    /// # Errors
    /// An appropriate [`CompilerError`] is returned if the compiler
    /// cannot handle the query.
    fn compile_query(query: &JsonPathQuery) -> Result<Self::E<'_>, CompilerError>;

    /// Turn a compiled [`Automaton`] into an [`Engine`].
    fn from_compiled_query(automaton: Automaton<'_>) -> Self::E<'_>;
}
