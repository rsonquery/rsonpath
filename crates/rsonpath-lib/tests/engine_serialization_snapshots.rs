#[cfg(feature = "serde")]
mod ron {
    use insta::assert_ron_snapshot;
    use rsonpath::engine::{Compiler, RsonpathEngine};
    use std::error::Error;

    fn engine(string: &str) -> Result<RsonpathEngine, Box<dyn Error>> {
        let query = rsonpath_syntax::parse(string)?;
        let engine = RsonpathEngine::compile_query(&query)?;
        Ok(engine)
    }

    #[test]
    fn empty_query() -> Result<(), Box<dyn Error>> {
        assert_ron_snapshot!(&engine("$")?);
        Ok(())
    }

    #[test]
    fn readme_query() -> Result<(), Box<dyn Error>> {
        assert_ron_snapshot!(&engine("$.jsonpath[*]")?);
        Ok(())
    }

    #[test]
    fn jsonpath_example_query() -> Result<(), Box<dyn Error>> {
        assert_ron_snapshot!(&engine("$..phoneNumbers[*].number")?);
        Ok(())
    }

    #[test]
    fn real_life_query() -> Result<(), Box<dyn Error>> {
        assert_ron_snapshot!(&engine("$.personal.details.contact.information.phones.home")?);
        Ok(())
    }

    #[test]
    fn slice() -> Result<(), Box<dyn Error>> {
        assert_ron_snapshot!(&engine("$..entries[3:5:7]")?);
        Ok(())
    }
}
