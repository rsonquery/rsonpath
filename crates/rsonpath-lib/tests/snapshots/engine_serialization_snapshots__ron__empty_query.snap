---
source: crates/rsonpath-lib/tests/engine_serialization_snapshots.rs
expression: "&engine(\"$\")?"
snapshot_kind: text
---
(V1, Automaton(
  states: [
    StateTable(
      attributes: StateAttributes(2),
      member_transitions: [],
      array_transitions: [],
      fallback_state: State(0),
    ),
    StateTable(
      attributes: StateAttributes(1),
      member_transitions: [],
      array_transitions: [],
      fallback_state: State(0),
    ),
  ],
))
